Introduce usage of `__builtin_unreachable` and `noreturn`

JohnoKing opened the pull request:
This pull request adds usage of the `__builtin_unreachable()` compiler builtin and the `noreturn` attribute. The purpose of these is to improve GCC/Clang compiler optimizations (this change also fixes a few compiler warnings). These are used when a function doesn't return (such as `errormsg` used with `ERROR_exit(1)`). Changes of note:

src/cmd/ksh93/sh/io.c: `outexcept()`:
\- Avoid using `__builtin_unreachable()` here since `errormsg` can return despite using `ERROR_system(1)` (see https://github.com/att/ast/issues/1336).

src/cmd/ksh93/tests/io.sh:
\- Add a regression test for the ksh2020 bug referenced above.

src/lib/libast/features/common:
\- Detect the existence of either the C11 `stdnoreturn.h` header or the GCC `noreturn` attribute, preferring the former when available.
\- Test for the existence of `__builtin_unreachable()` (the feature test is from https://github.com/ksh93/ksh/pull/243#issuecomment-808569808).

JohnoKing committed:
Introduce usage of __builtin_unreachable and noreturn

This commit adds usage of the __builtin_unreachable() compiler builtin
and the noreturn attribute. The purpose of these is to improve GCC/Clang
compiler optimizations (this change also fixes a few compiler warnings).
These are used when a function doesn't return (such as errormsg used
with ERROR_exit(1)). Changes of note:

src/cmd/ksh93/sh/io.c: outexcept():
- Avoid using __builtin_unreachable() here since errormsg can return
  despite using ERROR_system(1) (see https://github.com/att/ast/issues/1336).

src/cmd/ksh93/tests/io.sh:
- Add a regression test for the ksh2020 bug referenced above.

src/lib/libast/features/common:
- Detect the existence of either the C11 stdnoreturn.h header or the GCC
  noreturn attribute, preferring the former when available.
- Test for the existence of __builtin_unreachable() (the feature test is
  from https://github.com/ksh93/ksh/pull/243#issuecomment-808569808).

JohnoKing committed:
Add a missing UNREACHABLE() after sh_syntax

JohnoKing committed:
Merge branch 'master' of https://github.com/ksh93/ksh into noreturn-optimizations

JohnoKing committed:
Remove noreturn attribute from path_exec

One of the two path_exec calls (in b_exec) is followed by a call to
sh_done. In light of <https://github.com/att/ast/issues/1336> I'll avoid
marking path_exec as noreturn.

McDutchie commented on src/cmd/ksh93/edit/vi.c:
```diff
@@ -1060,8 +1060,7 @@ static int cntlmode(Vi_t *vp)
 	}
-/* NOTREACHED */
-	return(0);
+	UNREACHABLE();
```
Either the original `/* NOTREACHED */` comment was wrong, or lines 1060-1061 are dead code. Seems to be that can be reached if the `ed_getchar()` encounters EOF...

JohnoKing committed:
edit/vi.c: Revert return(0) -> UNREACHABLE() change

JohnoKing committed:
Merge branch 'master' into noreturn-optimizations

McDutchie committed:
__builtin_unreachable() for release, abort() for dev builds

src/lib/libast/features/common:
- Use __builtin_unreachable() for release builds, abort() for
  development builds. abort() aborts the program reliably and
  produces a stack trace, aiding in debugging.
  __builtin_unreachable() may or may not do the same -- its
  behaviour is undefined. So we shouldn't use it for that.

McDutchie committed:
io.c: add comment to clarify what's going on

McDutchie committed:
declare path_exec() noreturn; rm superflous sh_done() from b_exec()

McDutchie committed:
path.h: add forgotten noreturn (re: ba3f1a60)

McDutchie committed:
consistency: rm UNREACHABLE() calls after noreturn sh_syntax(), sh_done()
